### PR TITLE
KTOR-4318, KTOR-7561: Fix digest auth parameters

### PR DESCRIPTION
**Pull request by osipxd**
**Subsystem**
Client/Server, 

**Motivation**
- [KTOR-4318](https://youtrack.jetbrains.com/issue/KTOR-4318) Digest Auth does not implement nc parameter correctly according to RFC 7616
- [KTOR-7561](https://youtrack.jetbrains.com/issue/KTOR-7561) Digest auth: `username` and `cnonce` parameters aren't surrounded with quotes

**Solution**
Update parameters’ implementation to follow RFC 7616 guidelines.

This PR fixes only these two bugs, while we have more implementation inconsistencies with the RFC. I've created a separate issue to update Digest Auth implementation according to the RFC - [KTOR-7578](https://youtrack.jetbrains.com/issue/KTOR-7578/Update-Digest-authentication-implementation-according-to-RFC-7616)

It's better to review commit-by-commit

### CODE REVIEW

File: ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt [hunk 1]
```diff
@@ -25,3 +25,7 @@ public enum class HeaderValueEncoding {
      */
     URI_ENCODE
 }
+
+public fun interface HeaderValueEncodingStrategy {
+    public fun chooseEncoding(name: String): HeaderValueEncoding
+}
```
Comment (osipxd) [inline ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt:31]: If this solution is OK, I will add documentation for this interface.

File: ktor-client/ktor-client-plugins/ktor-client-auth/api/ktor-client-auth.api [hunk 2]
```diff
@@ -113,6 +113,7 @@ public final class io/ktor/client/plugins/auth/providers/DigestAuthCredentials {
 }
 
 public final class io/ktor/client/plugins/auth/providers/DigestAuthProvider : io/ktor/client/plugins/auth/AuthProvider {
+	public static final field Companion Lio/ktor/client/plugins/auth/providers/DigestAuthProvider$Companion;
```
Comment (e5l) [inline ktor-client/ktor-client-plugins/ktor-client-auth/api/ktor-client-auth.api:116]: Should we make the companion internal?

File: ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt (hunk 1 above)
Comment (e5l) [reply ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt:31]: It looks like the target branch should be at least 3.1.0-eap, please check if it is possible to fix without new public api

**Review by e5l — APPROVED** [review]

**Comment by e5l** [comment]
lgtm, please check if it can be fixed in patch

File: ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt (hunk 1 above)
Comment (e5l) [reply ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt:31]: It also might be good to encode directly, so strategy will be an edge case

**Review by e5l — CHANGES_REQUESTED** [review]
lgtm, left some comments

File: ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt (hunk 1 above)
Comment (osipxd) [reply ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt:31]: Do you mean adding some `HeaderValueEncoder` which will be responsible for encoding header values?

Comment (e5l) [reply ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt:31]: Yep

Comment (osipxd) [reply ktor-http/common/src/io/ktor/http/auth/HeaderValueEncoding.kt]: I've managed to fix these bugs without introducing new APIs

**Review by e5l — APPROVED** [review]
Nice, lgtm! 
Please add issue numbers in the issue title :)

**Pull request merged**